Spark 3.2: Implement merge-on-read UPDATE

aokolnychyi opened the pull request:
This PR implements merge-on-read UPDATE in Spark 3.2.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/io/BasePositionDeltaWriter.java:
```diff
@@ -27,25 +27,35 @@
 public class BasePositionDeltaWriter<T> implements PositionDeltaWriter<T> {
 
-  private final PartitioningWriter<T, DataWriteResult> dataWriter;
+  private final PartitioningWriter<T, DataWriteResult> insertWriter;
```
Let's discuss the interface change instead (below).

aokolnychyi commented on core/src/main/java/org/apache/iceberg/io/PositionDeltaWriter.java:
```diff
@@ -41,6 +41,21 @@
+   * @param spec a new partition spec
+   * @param partition a new partition or null if the spec is unpartitioned
+   */
+  default void update(T row, PartitionSpec spec, StructLike partition) {
```
This needs to be discussed and I am open to alternatives. With the current change, the API seems a bit confusing but maybe we can play with names to make the purpose clear. In short, this method is to distinguish inserts from updates. In case of MERGE, we order updates by _spec_id, _partition, _file, _pos and updated records can be out of order if we change the sort key or partition columns. For inserts, the metadata columns are null so we have no problems ordering inserts.

aokolnychyi replied:
Splitting an update into two records (separate delete and separate insert) on the engine side is an option to consider but will require quite some effort to implement. Ideally, I would avoid that.

rdblue replied:
If I understand correctly, the problem that this fixes is that the inserts are expected to be clustered, but we don't have the same expectation for updates because it's hard to come up with a single sort order that works for all cases (inserts, updates, and deletes).

So we separate updates that are not expected to be clustered and use a fanout writer for them?

szehon-ho replied:
What do you think to have update take arguments (CharSequence path, long pos, T row, PartitionSpec spec, StructLike partition) and internally invoke delete()?  Maybe more fitting of the name, than having the caller call delete() first and then this method.

rdblue replied:
Actually, both insert and update writers appear to be clustered. Do they just have a different clustering?

rdblue replied:
I see. The partitioned writer has a comment about this. Updates come in so that deletes are clustered and ordered correctly -- by the path and position that will be deleted. But updates may have caused the row to change partitions and so that clustering for deletes may not match clustering for updates. So we do need to use a separate fanout writer.

aokolnychyi replied:
@szehon-ho, I thought about it but the original record may be in one spec/partition and the new record may be written to another one. As a result, we would need to have path, pos, row, oldSpec, oldPartition, newSpec, newPartition, which makes it a rather long list of arguments. I am happy to reconsider if folks prefer that API instead of this one.

@rdblue, correct, your last comment is spot on.

szehon-ho replied:
Yea I see, that is unfortunately a bit of arguments, though still less than max limit of checkstyle.  Yea in my opinion, one single method is much cleaner than having the client remembering to call this one after delete(), but it is definitely a style choice and also not sure if anyone else agrees (@rdblue , @RussellSpitzer @jackye1995 ). Else we might  rename the method to something a bit awkward like 'updateDeleted' to be more clear.

aokolnychyi replied:
Let me create an issue where we can discuss these two options to avoid blocking the subsequent MERGE PR.

aokolnychyi replied:
Created #4042.

szehon-ho replied:
Sure no problem, don't let it block

aokolnychyi commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:
```diff
@@ -198,7 +198,7 @@ private DistributionMode adjustWriteDistributionMode(DistributionMode mode) {
     }
   }
 
-  public DistributionMode copyOnWriteDeleteDistributionMode() {
```
Given the recent changes, I am no longer sure we need separate methods for copy-on-write and merge-on-read.

aokolnychyi commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:
```diff
@@ -407,6 +428,216 @@ public void close() throws IOException {
     }
   }
 
+  private static class UnpartitionedDeltaWriter extends BaseDeltaWriter {
```
Both `UnpartitionedDeltaWriter` and `PartitionedDeltaWriter` share the `delete` implementation but `insert` and `update` methods are different. We could refactor these two into a single classs but that would raise questions about extra method calls. We could go with an accessor and hope it will be inlined by JIT but we can't know for sure.

aokolnychyi commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:
```diff
@@ -1441,6 +1441,132 @@ public void testRangePositionDeltaDeletePartitionedTable() {
+  // Distribution and ordering for merge-on-read UPDATE operations with position deletes
+  // ===================================================================================
+  //
+  // update mode is NOT SET -> CLUSTER BY _spec_id, _partition + LOCALLY ORDER BY _spec_id, _partition, _file, _pos
```
Same like in merge-on-read DELETE. Here we are dealing with full rows so it does not seem like a good idea to group everything into a single task for unpartitioned tables. Ideas are welcome.

rdblue replied:
I think this is a reasonable default. We don't expect high volume for merge-on-read, nor do we expect unpartitioned tables to be high volume. If they are, then setting an explicit mode=none will fix it.

aokolnychyi replied:
Okay, let's keep it for now then.

rdblue commented on data/src/test/java/org/apache/iceberg/io/TestPositionDeltaWriters.java:
```diff
@@ -79,13 +79,16 @@ public void setupTable() throws Exception {
-    ClusteredDataWriter<T> dataWriter = new ClusteredDataWriter<>(
+    ClusteredDataWriter<T> insertWriter = new ClusteredDataWriter<>(
+        writerFactory, fileFactory, table.io(),
+        fileFormat, TARGET_FILE_SIZE);
```
Does this need to wrap here?

aokolnychyi replied:
Wrapped for the consistency with the rest of the code around.

rdblue commented on data/src/test/java/org/apache/iceberg/io/TestPositionDeltaWriters.java:
```diff
@@ -144,13 +147,16 @@ public void testPositionDeltaDeleteOnly() throws IOException {
-    ClusteredDataWriter<T> dataWriter = new ClusteredDataWriter<>(
+    ClusteredDataWriter<T> insertWriter = new ClusteredDataWriter<>(
+        writerFactory, fileFactory, table.io(),
+        fileFormat, TARGET_FILE_SIZE);
```
Nit: Unnecessary line wrap?

aokolnychyi replied:
Same here.

rdblue commented on spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:
```diff
@@ -126,6 +128,33 @@ object RewriteUpdateTable extends RewriteRowLevelCommand {
+    val project = Project(operationType +: updatedRowsPlan.output, updatedRowsPlan)
+
+    // build a plan to write the row delta to the table
+    val writeRelation = relation.copy(table = table)
```
This caught me the first time I saw it, too. What's happening here is that we're replacing the original table with the `RowLevelOperationTable` that wraps the `RowLevelOperation`. The reason is so that scans produced by the table are coordinated with writes done in the table. We use the table interface to avoid needing to handle the operation separately.

szehon-ho replied:
Ah sorry, tricked by variable shadowing..

aokolnychyi replied:
I'll try to change the name to make it a bit more obvious. I'll do that in a separate PR to cover all places at once.

rdblue commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:
```diff
@@ -407,6 +428,216 @@ public void close() throws IOException {
+    private final PositionDeltaWriter<InternalRow> delegate;
+    private final FileIO io;
+    private final PartitionSpec dataSpec;
+    private final Map<Integer, PartitionSpec> specs;
```
Why have `dataSpec` and `specs` for unpartitioned tables? Is this that this may be reading previously partitioned data?

aokolnychyi replied:
Correct, we need `specs` for delete projections and we need `dataSpec` to attach correct spec ID to the produced data file (since we did not reserve a particular spec ID for unpartitioned specs).

rdblue commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:
```diff
@@ -407,6 +428,216 @@ public void close() throws IOException {
+            context.dataFileFormat(), context.targetDataFileSize());
+      }
+
+      // always use a separate fanout data writer for updates as they may be out of order
```
Got it. This is what I was looking for.

rdblue replied:
If I remember correctly, this is because we sort by partition, file, and pos, but the correct output order depends on the new data. So we always delete with clustering, but the updated rows may not still conform to the original clustering. Right?

aokolnychyi replied:
Yeah, happens when we update a column that is used in the spec and the updated record moves to a different partition.

rdblue commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:
```diff
@@ -407,6 +428,216 @@ public void close() throws IOException {
+    }
+  }
+
+  private static class PartitionedDeltaWriter extends BaseDeltaWriter {
```
It definitely looks to me like there is a lot of duplication between the partitioned and unpartitioned cases. How bad is it to merge the two as you commented above? I think I would probably prefer that.

aokolnychyi replied:
I combined these two classes in one and exposed a protected field in the parent instead.

rdblue reviewed (approved):
Mostly looks good to me. A couple of minor things:
* Where is the case where an update repartitions a record (requiring the separate fanout writer) tested?
* Are there tests to ensure predicate pushdown and dynamic pruning work?
* There's a lot of duplication between the unpartitioned and partitioned writers. I'd prefer to fix that.

singhpk234 commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkDistributionAndOrderingUtil.java:
```diff
@@ -146,14 +146,14 @@ private static Distribution buildCopyOnWriteDeleteUpdateDistribution(Table table
+    if (command == DELETE || command == UPDATE) {
+      return buildPositionDeleteUpdateDistribution(distributionMode);
     } else {
       throw new IllegalArgumentException("Only position deletes are currently supported");
```
[minor] can update the exception message here as well, to include `update` command: 

```suggestion
      throw new IllegalArgumentException("Only position deletes and updates are currently supported");
```

aokolnychyi replied:
Good catch, let me fix it.

aokolnychyi commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:
```diff
@@ -407,6 +428,216 @@ public void close() throws IOException {
+    @Override
+    public void update(InternalRow meta, InternalRow id, InternalRow row) throws IOException {
+      delete(meta, id);
+      delegate.update(row, dataSpec, null);
```
Yes, the produced `DataFile` should have the correct spec ID.

aokolnychyi commented:
@rdblue, I've introduced a parent class to reduce the code duplication in unpartitioned and partitioned delta writers, added tests for out of order records when we update a partition column and added a test for static partition pruning. Dynamic pruning for merge-on-read is a little bit tricky to trigger (we just use the regular scan). I'll give it another try in MERGE.

aokolnychyi commented:
Thanks for reviewing, @szehon-ho @rdblue @singhpk234! Issue #4042 remains open.
